src/ber/parser.rs

```diff
@@ -19,14 +19,29 @@ pub(crate) fn bytes_to_u64(s: &[u8]) -> Result<u64, BerError> {
     Ok(u)
 }
 
-pub(crate) fn parse_identifier(i: &[u8]) -> IResult<&[u8], (u8, u8, u8)> {
+pub(crate) fn parse_identifier(i: &[u8]) -> IResult<&[u8], (u8, u8, u32)> {
     if i.is_empty() {
         Err(Err::Incomplete(Needed::Size(1)))
     } else {
         let a = i[0] >> 6;
         let b = if i[0] & 0b0010_0000 != 0 { 1 } else { 0 };
-        let c = i[0] & 0b0001_1111;
-        Ok((&i[1..], (a, b, c)))
+        let mut c = (i[0] & 0b0001_1111) as u32;
+
+        let mut tag_byte_count = 1;
+
+        if c == 0x1f {
+            c = 0;
+            loop {
+                c = (c << 7) | ((i[tag_byte_count] as u32) & 0x7f);
```

Comment:
This has two problems:
- `tag_byte_count` is not tested to be in bounds, so it can panic
- the left shift can overflow
It may be easy to work by reducing the slice, after testing it for size. Example using the `error_if` macro:
```rust
error_if!(i, i.is_empty(), ErrorKind::Custom(BER_TAG_ERROR))?;
let i = &i[1..];
```

For the overflow, it needs to be checked at each iteration, before shifting:
```rust
error_if!(i, c & 0xff00_0000 != 0, ErrorKind::Custom(BER_INVALID_LENGTH))?;
c = (c << 7) | ((i[0] as u32) & 0x7f);
```

Reply:
Thanks! Should be sorted in the commit I just pushed.